README.md
# mp3hash

[![Build Status](https://travis-ci.org/jvrsantacruz/mp3hash.png?branch=master)](https://travis-ci.org/jvrsantacruz/mp3hash)

Hashes music files ignoring meta-data.

Useful to detect the same song in different tagged files.

The following meta-data standards are supported:

    id3v1, id3v1 extended, id3v2.2, id3v2.3 and id3v.24

Known to work with: `python2.6` and `python2.7`

Javier Santacruz (2012-2013)

# Command line tool

Similarly to other well-known tools like `sha1sum` or `md5sum`, it takes one or more files and
outputs the hashes. It matches those program's interface as much as possible.

```bash
$ mp3hash *.mp3
6611bc5b01a2fc6a6386a871e8c51f86e1f12b33 13_Hotel-California-(Gipsy-Kings).mp3
6611bc5b01a2fc6a6386a871e8c51f86e1f12b33 14_Hotel-California-(Gipsy-Kings).mp3
```

You can see how it returns the same hash number for either file, even though their tags are
different, and so the data inside them. 

```bash
$ sha1sum *.mp3
6a1d5f8317add10e205ae30174630b47645fb5b4  13_Hotel-California-(Gipsy-Kings).mp3
c28d6976114d31df3366d9935eb0bedd36cf1f0b  14_Hotel-California-(Gipsy-Kings).mp3
```

The hash it's made strictly using the music data in the file. This is done by calculating the tags
sizes and ignoring them when calculating the hash, thus hashing only the music data in the file.

The default hashing algorithm is `sha-1`, but any algorithm can be used as long it's supported by
the Python's `hashlib` module. A complete list of all available hashing algorithms can be obtained
by calling the program with the `--list-algorithms` flag.

```bash
$ ./mp3hash --list-algorithms
md5
sha1
sha224
sha256
sha384
sha512
```

Any of the algorithms listed will be available to consume and hash the content files.

```bash
./mp3hash --algorithm md5
ac0fdd89454528d3fbdb19942a2e6653 13_Hotel-California-(Gipsy-Kings).mp3
ac0fdd89454528d3fbdb19942a2e6653 14_Hotel-California-(Gipsy-Kings).mp3
```

You can even extend the library with your own hash functions, see the _development_ section to read
about the API and how to use it.

# Installation

It doesn't have any dependences besides `python2.7+`.

In order to access to the `mp3hash` script, the package should be installed.

    python setup.py install

Just afterwards, the `mp3hash` command should be available in path.

# API

The main components are the `mp3hash` function and the `TaggedFile` class.

## mp3hash

`mp3hash.mp3hash` will compute the hash on the music (and only the music)
  of the file in the given path.

```python
>> from mp3hash import mp3hash
>> mp3hash('/path/to/song.mp3')
Out: 6611bc5b01a2fc6a6386a871e8c51f86e1f12b33
```

## TaggedFile

`mp3hash.TaggedFile` class takes a file-like object supporting seek with negative values and will
parse all the sizes and offsets for the meta-data tags stored within it.

```python
>> from mp3hash import TaggedFile
>> with open('/path/to/song.mp3') as file:
    TaggedFile(file).has_id3v1
Out: True

    TaggedFile(file).has_id3v2
Out: True

    TaggedFile(file).filesize
Out: 5315937

    TaggedFile(file).music_size
Out: 5311714

    TaggedFile(file).startbyte
Out: 4096

    TaggedFile(file).music_limits
Out: (4096, 5315810)
```

## Bring your own hash/checksum!

Any object matching the `update` and `hexdigest` methods, follows the hasher protocol and thereby
can be used along with the `mp3hash` function.

If your method happens to not to match this protocol, you can always adapt it. We could carry out a
little experiment. It should be easy enough for us to wrap the much faster `adler32` checksum
algorithm to make it work with `mp3hash`.

The algorithm is available in the python standard module `zlib`, as `zlib.adler32`.

From the documentation:

> zlib.adler32(data[, value])
>
> Computes a Adler-32 checksum of data. [..] If value is present, it is used as the starting value
	of the checksum; otherwise, a fixed default value is used. This allows computing a running
	checksum over the concatenation of several inputs. [..]

```python
>> import zlib
>> import mp3hash

>> class Adler32Hasher(object):
      def __init__(self):
          self.value = None

      def update(self, data):
      	  if self.value is None:  # first call
      	      self.value = zlib.adler32(data)
      	  else:
      	      self.value = zlib.adler32(data, self.value)

      def hexdigest(self):
          return hex(self.value)

>> mp3hash.mp3hash('/path/to/song.mp3', hasher=Adler32Hasher())
Out: '0x40b1519d'
```

# Developers, developers, developers!

## Testing environment

You're encouraged to use a _virtualenv_

```bash
$ virtualenv --python python2 --distribute env
$ source env/bin/activate
```

Once into the _virtualenv_, install the package and the testing dependences.

```bash
$(env) python setup.py develop
$(env) pip install -r dev-reqs.txt
```

In order to perform the testing, use the `nosetests` test runner and collector from the root of the
project (same directory as of the `setup.py` file).

```bash
$ nosetests
```

## About id3v1

- id3v1 is 128 bytes at the end of the file starting with 'TAG'
- id3v1 extended is 227 bytes before regular id3v1 tag starting with 'TAG+'

total size: 128 + (227 if extended)

Based on id3v1 wikipedia docs:

- http://en.wikipedia.org/wiki/ID3


## About id3v2

- id3v2 has a 10 bytes header at the begining of the file.
	- byte 5 holds flags. 4th bit indicates presence of footer in v2.4
   	- bytes 6-10 are the tag size (not counting header)


total size: header + tagsize + footer (if any)

Based on id3v2 docs:

- http://id3.org/id3v2-00
- http://id3.org/id3v2.3.0
- http://id3.org/id3v2.4.0-structure


CHANGES.txt
0.1 (2013-04-15)
------------------

* Adds support for python 2.6
* Removes mutagen from development requirements
* Changes hash function signature to accept any object matching hasher protocol.
* Tests for the command line interface
* Adds support for tox environment testing

0.0.6 (2013-04-10)
------------------

* Adds id3v2.4 support
* Removes useless id3v2 extended tag parsing
* Improves documentation and tests

0.0.5 (2013-04-07)
------------------

* Fixes id3v2 size calculation
* Fixes gruesome integration tests
* Changes TaggedFile interface to use file-objects instead


0.0.4 (2013-03-17)
------------------

* Adds packaging and uploads it to pypi
* Refactor of the package splitting it into a library and a script


0.0.3 (2012-11-15)
------------------

* First integration tests


0.0.2 (2012-09-20)
------------------

* Adds public function mp3hash


0.0.1 (2012-06-03) 
------------------

* First version


tests/test_hash.py
#-*- coding: utf-8 -*-

import os
import shutil
import hashlib

import mutagen

from hamcrest import *
from nose.tools import raises

from mp3hash import mp3hash


SONG1_PATH = 'tests/file1.mp3'
SONG2_PATH = 'tests/file2.mp3'

SONG_SIZE = os.path.getsize(SONG1_PATH)


class IdenticalFiles(object):
    @classmethod
    def setup_class(cls):
        shutil.copy(SONG1_PATH, SONG2_PATH)


class TestSameDataButNoTags(IdenticalFiles):
    @classmethod
    def setup_class(cls):
        shutil.copy(SONG1_PATH, SONG2_PATH)
        f = mutagen.File(SONG2_PATH)
        f.clear()
        f.save()


class HashOperations(object):
    def test_mp3hash(self):
        hash1 = mp3hash(SONG1_PATH)
        hash2 = mp3hash(SONG2_PATH)
        assert_that(hash1, is_(equal_to(hash2)))

    def test_algs(self):
        "Test generator for every algorithm"
        for alg in hashlib.algorithms:
            yield self.check_algs, alg

    def check_algs(self, alg):
        hash1 = mp3hash(SONG1_PATH, alg=alg)
        hash2 = mp3hash(SONG2_PATH, alg=alg)
        assert_that(hash1, is_(equal_to(hash2)))

    def test_maxbytes_all(self):
        "Test generator for multiple sizes"
        for num in range(1, SONG_SIZE, 250 * 1024):
            yield self.check_num, num

    def check_num(self, maxbytes):
        hash1 = mp3hash(SONG1_PATH, maxbytes=maxbytes)
        hash2 = mp3hash(SONG2_PATH, maxbytes=maxbytes)
        assert_that(hash1, is_(equal_to(hash2)))

    @raises(ValueError)
    def test_maxbytes_negative(self):
        mp3hash(SONG1_PATH, maxbytes=-15)

    @raises(ValueError)
    def test_maxbytes_0(self):
        mp3hash(SONG1_PATH, maxbytes=-15)


class TestHashIdenticFile(IdenticalFiles, HashOperations):
    pass


class TestHashSameDataButNoTags(TestSameDataButNoTags, HashOperations):
    pass


tests/unit/test_id3v1.py
#-*- coding: utf-8 -*-

from cStringIO import StringIO

from hamcrest import assert_that, is_

from mp3hash import TaggedFile, ID3V1_SIZE


MP3_ID3v1_NOT_TAGGED = '\n' * ID3V1_SIZE
MP3_ID3v1_TAGGED = 'TAG' + '\n' * (ID3V1_SIZE - 3)
MP3_ID3v1_TAGGED_AND_FILLED = '\n' * 512 + MP3_ID3v1_TAGGED


class TestID3v1(object):
    def test_detects_id3v1_tags(self):
        file = StringIO(MP3_ID3v1_TAGGED)

        tagged = TaggedFile(file)

        assert_that(tagged.has_id3v1)

    def test_detects_id3v1_tags_even_with_content(self):
        file = StringIO(MP3_ID3v1_TAGGED_AND_FILLED)

        tagged = TaggedFile(file)

        assert_that(tagged.has_id3v1)

    def test_detects_when_there_is_no_id3v1_tag(self):
        file = StringIO(MP3_ID3v1_NOT_TAGGED)

        tagged = TaggedFile(file)

        assert_that(not tagged.has_id3v1)

    def test_says_there_is_no_tag_when_file_is_too_small(self):
        file = StringIO()

        tagged = TaggedFile(file)

        assert_that(not tagged.has_id3v1)


class TestID3v1Size(object):
    def test_id3v1_size_is_128_if_there_is_tag(self):
        file = StringIO(MP3_ID3v1_TAGGED)

        tagged = TaggedFile(file)

        assert_that(tagged.id3v1_size, is_(ID3V1_SIZE))

    def test_id3v1_size_is_0_if_no_tag_is_present(self):
        file = StringIO(MP3_ID3v1_NOT_TAGGED)

        tagged = TaggedFile(file)

        assert_that(tagged.id3v1_size, is_(0))


tests/unit/test_id3v2.py
#-*- coding: utf-8 -*-

from cStringIO import StringIO

from hamcrest import assert_that, is_

from mp3hash import TaggedFile, ID3V2_HEADER_SIZE, ID3V2_FOOTER_SIZE


NOT_TAGGED = '\n' * ID3V2_HEADER_SIZE
TAGGED = 'ID3' + '\n' * (ID3V2_HEADER_SIZE - 3)
TAGGED_AND_FILLED = TAGGED + '\n' * 512


class TestID3v2(object):
    def test_detects_id3v2_tags(self):
        file = StringIO(TAGGED)

        tagged = TaggedFile(file)

        assert_that(tagged.has_id3v2)

    def test_detects_id3v2_tags_even_with_content(self):
        file = StringIO(TAGGED_AND_FILLED)

        tagged = TaggedFile(file)

        assert_that(tagged.has_id3v2)

    def test_detects_when_there_is_no_id3v1_tag(self):
        file = StringIO(NOT_TAGGED)

        tagged = TaggedFile(file)

        assert_that(not tagged.has_id3v2)

    def test_says_there_is_no_tag_when_file_is_too_small(self):
        file = StringIO()

        tagged = TaggedFile(file)

        assert_that(not tagged.has_id3v2)


VERSION = 0x03
REVISION = 0x0
FLAGS = 0x0
RAW_SIZE = [0x0, 0x0, 0x02, 0x01]  # (2 << 7) + 1 = 257
SIZE = 257

PARSED_HEADER = ('ID3', VERSION, REVISION, FLAGS, SIZE)
RAW_HEADER = [VERSION, REVISION, FLAGS] + RAW_SIZE
HEADER = 'ID3' + ''.join(map(chr, RAW_HEADER))


class TestID3v2Sizes(object):
    def test_header_is_read_properly(self):
        file = StringIO(HEADER)

        tagged = TaggedFile(file)

        assert_that(tagged._id3v2_header, is_(PARSED_HEADER))

    def test_size_is_read_from_header_correctly(self):
        file = StringIO(HEADER)

        tagged = TaggedFile(file)

        assert_that(tagged.id3v2_size, is_(SIZE + ID3V2_HEADER_SIZE))

    def test_id3v2_totalsize_is_id3v2_size(self):
        assert_that(TaggedFile.id3v2_totalsize, is_(TaggedFile.id3v2_size))


VERSION_v24 = 0x4
FLAGS_v24 = 0x10  # footer present set

PARSED_HEADER_v24 = ('ID3', VERSION_v24, REVISION, FLAGS_v24, SIZE)
RAW_HEADER_v24 = [VERSION_v24, REVISION, FLAGS_v24] + RAW_SIZE
HEADER_v24 = 'ID3' + ''.join(map(chr, RAW_HEADER_v24))


class TestID3v24Sizes(object):
    def test_header_is_read_properly(self):
        file = StringIO(HEADER_v24)

        tagged = TaggedFile(file)

        assert_that(tagged._id3v2_header, is_(PARSED_HEADER_v24))

    def test_size_adds_footer_size(self):
        file = StringIO(HEADER_v24)

        tagged = TaggedFile(file)

        assert_that(tagged.id3v2_size, is_(SIZE + ID3V2_HEADER_SIZE + ID3V2_FOOTER_SIZE))


cli.py
#!/usr/bin/env python
#-*- coding: utf-8 -*-

import os
import sys
import errno
import hashlib
from optparse import OptionParser

from mp3hash import mp3hash


# hashlib.algorithms was included in python 2.7
ALGORITHMS = getattr(
    hashlib,
    'algorithms',
    ('md5', 'sha1', 'sha224', 'sha256', 'sha384', 'sha512')
)


def error(msg, code=1):
    print(msg)
    sys.exit(code)


def list_algorithms():
    print(u'\n'.join(ALGORITHMS))


def main():
    opts, args, parser = parse_arguments()

    if not args and not opts.list_algorithms:
        parser.print_help()
        error(u"\nInsufficient arguments", code=errno.EINVAL)

    if opts.maxbytes is not None and opts.maxbytes <= 0:
        parser.print_help()
        print()
        error(u"\nInvalid value for --maxbytes it should be a positive integer",
              code=errno.EINVAL)

    if opts.list_algorithms:
        list_algorithms()
        sys.exit()

    if opts.algorithm not in ALGORITHMS:
        error(u"Unkown '{}' algorithm. Available options are: {}"
              .format(opts.algorithm, ", ".join(ALGORITHMS)),
              code=errno.EINVAL)

    for arg in args:
        path = os.path.realpath(arg)
        if not os.path.isfile(path):
            error(u"File at '{}' does not exist or it is not a regular file"
                  .format(arg), code=errno.ENOENT)
            continue

        hasher = hashlib.new(opts.algorithm)

        print(u'{hash} {filename}'.format(
            hash=mp3hash(path, maxbytes=opts.maxbytes, hasher=hasher),
            filename=os.path.basename(path) if not opts.hash else ''
        ))


def parse_arguments():
    parser = OptionParser()

    parser.add_option("-a", "--algorithm", default='sha1',
                      help="Hash algorithm to use. Default sha1.  "
                      "See --list-algorithms")

    parser.add_option("-l", "--list-algorithms", action="store_true",
                      default=False, help="List available algorithms")

    parser.add_option("-q", "--hash", action="store_true", default=False,
                      help="Print only hash information, no filename")

    parser.add_option("-m", "--maxbytes", type=int, default=None,
                      help="Max number of bytes of music to hash")

    parser.add_option("-o", "--output", default=False,
                      help="Redirect output to a file")

    parser.set_usage("Usage: [options] FILE [FILE ..]")

    (opts, args) = parser.parse_args()

    return opts, args, parser


def configure_output(output):
    if output:
        stdout = sys.stdout
        try:
            sys.stdout = open(output, 'w')
        except IOError as err:
            sys.stdout = stdout
            error(u"Couldn't open {}: {}".format(output, err), errno.ENOENT)


if __name__ == "__main__":
    main()


mp3hash.py
#-*- coding: utf-8 -*-
"""
Hashes audio files ignoring metadata

The following metadata standards are supported:

    id3v1, id3v1 extended, id3v2.2, id3v2.3 and id3v.24

The main components are the mp3hash function and the TaggedFile class.

* mp3hash will compute the hash on the music (and only the music)
  of the file in the given path.

* TaggedFile class takes a file-like object supporting
  seek and negative values for seek and will parse all the sizes
  for the metadata stored within it.

Technical details:
~~~~~~~~~~~~~~~~~~

id3v1 is 128 bytes at the end of the file starting with 'TAG'
id3v1 extended is 227 bytes before regular id3v1 tag starting with 'TAG+'

total size: 128 + (227 if extended)

id3v2 has a 10 bytes header at the begining of the file.
      byte 5 holds flags. 4th bit indicates presence of footer in v2.4
      bytes 6-10 are the tag size (not counting header)

total size: header + tagsize + footer (if any)

Based on id3v1 wikipedia docs: http://en.wikipedia.org/wiki/ID3
Based on id3v2 docs: http://www.id3.org/id3v2.3.0

Javier Santacruz 2012-06-03
"""

import struct
import hashlib
from collections import deque
from itertools import repeat, chain


def mp3hash(path, maxbytes=None, hasher=None):
    """Returns the hash of the sound contents of a ID3 tagged file
    Convenience function which wraps TaggedFile
    Returns None on failure
    """
    if maxbytes is not None and maxbytes <= 0:
        raise ValueError(u'maxbytes must be a positive integer')

    if hasher is None:
        hasher = hashlib.new('sha1')

    with open(path, 'rb') as ofile:
        return TaggedFile(ofile).hash(maxbytes=maxbytes, hasher=hasher)


def hashfile(file, start, end, hasher, maxbytes=None, blocksize=2 ** 19):
    """Hashes an open file data starting from byte 'start' to the byte 'end'
    max is the maximum amount of data to hash, in bytes.
    The hexdigest string is calculated considering only bytes between start,end
    default block size is 512 KiB
    """
    if maxbytes is not None and maxbytes > 0:
        end = min(end, start + maxbytes)

    read, update = file.read, hasher.update  # Operations

    size = end - start
    nblocks = size // blocksize
    spare_block_size = size % blocksize

    blocksizes = repeat(blocksize, nblocks)
    if spare_block_size:
        blocksizes = chain(blocksizes, [spare_block_size])

    file.seek(start)  # jump headers
    consume(update(read(size)) for size in blocksizes)

    return hasher.hexdigest()


def consume(iterator):
    """ Consume the entire iterator ignoring its result """
    # feed the entire iterator into a 0-length deque
    deque(iterator, maxlen=0)


def memento(function):
    def wrapper(self):
        attr_name = '_' + function.__name__ + '_value'
        if not hasattr(self, attr_name):
            setattr(self, attr_name, function(self))
        return getattr(self, attr_name)
    return wrapper


def parse_7bitint(bytes, bits=7, mask=(1 << 7) - 1):
    """ Parses a big endian integer from a list of bytes
    taking only the first 7 bits from each byte

    Do not set bits and mask parameters.

    mask: 0111 1111 (removes 8th bit)
    shift: offset to reorder bytes as if the 8th didn't exist
    the bytes are walked in reverse order, as it is in big endian

    This is because ID3v2 uses 'sync safe integers'
    which always have its mrb zeroed

    > The tag size is encoded with four bytes where the most
    significant bit (bit 7) is set to zero in every byte, making a total
    of 28 bits. The zeroed bits are ignored, so a 257 bytes long tag is
    represented as $00 00 02 01.
    """
    return sum(
        (bytes[-i - 1] & mask) << shift
        for i, shift in
        enumerate(range(0, len(bytes) * bits, bits))
    )


ID3V1_SIZE = 128
ID3V1_EXTENDED_SIZE = ID3V1_SIZE + 227

ID3V2_HEADER_SIZE = 10
ID3V2_FOOTER_SIZE = 10


class TaggedFile(object):

    def __init__(self, file):
        self.file = file
        self.file.seek(0, 2)  # end of file
        self.filesize = self.file.tell()

    @property
    @memento
    def has_id3v1(self):
        "Returns True if the file is id3v1 tagged"
        if self.filesize < ID3V1_SIZE:
            return False

        self.file.seek(-ID3V1_SIZE, 2)  # last bytes of file
        return self.file.read(3) == 'TAG'

    @property
    @memento
    def has_id3v1ext(self):
        "Returns True if the file is id3v1 with extended tag"
        if self.filesize < ID3V1_EXTENDED_SIZE:
            return False

        self.file.seek(-ID3V1_EXTENDED_SIZE, 2)  # 227 before regular tag
        return self.file.read(4) == 'TAG+'

    @property
    @memento
    def id3v1ext_size(self):
        "Returns the size of the extended tag if exists"
        return ID3V1_EXTENDED_SIZE if self.has_id3v1ext else 0

    @property
    @memento
    def id3v1_size(self):
        "Returns the size in bytes of the id3v1 tag"
        return ID3V1_SIZE if self.has_id3v1 else 0

    @property
    @memento
    def id3v1_totalsize(self):
        "Returns the size in bytes of the id3v1 tag"
        return self.id3v1_size + self.id3v1ext_size

    @property
    @memento
    def _id3v2_header(self):
        """Returns id3v2 header: (id3, version, revision, flags, size)

        id3v2 header is 10 bytes long which starts with ID3:

            ID3v2/file identifier      "ID3"
            ID3v2 version              $03 00
            ID3v2 flags                %abcd0000
            ID3v2 size             4 * %0xxxxxxx

        Flags:  The b (6th) bit indicates whether or not the header is
                followed by an extended header. (mask is 0x40)

        (v2.4)  The d (4th) bit indicates that a footer  is present at the
                end of the tag. (mask is 0x10)
        """
        self.file.seek(0)
        header = self.file.read(ID3V2_HEADER_SIZE)

        id3, v, r, flags, size = struct.unpack('>3sBBB4s', header)

        return id3, v, r, flags, parse_7bitint(map(ord, size))

    @property
    @memento
    def has_id3v2(self):
        "Returns True if the file is id3v2 tagged"
        if self.filesize < ID3V2_HEADER_SIZE:
            return False

        id3, ver, rev, flags, size = self._id3v2_header
        return id3 == 'ID3'

    @property
    @memento
    def id3v2_size(self):
        """ Returns the size in bytes of the whole id3v2 tag

        > The ID3v2 tag size is the size of the complete tag after
        unsychronisation, including padding, excluding the header but not
        excluding the extended header.
        """
        if not self.has_id3v2:
            return 0

        id3, ver, rev, flags, size = self._id3v2_header

        # id3v2.4 also includes an optional 10 bytes footer
        if ver == 4 and flags & 0x10:
            size += ID3V2_FOOTER_SIZE

        return ID3V2_HEADER_SIZE + size

    id3v2_totalsize = id3v2_size

    @property
    @memento
    def startbyte(self):
        "Returns the starting byte position of music data in the file"
        return self.id3v2_totalsize

    @property
    @memento
    def endbyte(self):
        "Returns the last byte position of music data in the file"
        if not self.has_id3v1:
            return self.filesize

        self.file.seek(-self.id3v1_totalsize, 2)
        return self.file.tell()

    @property
    @memento
    def music_limits(self):
        "Returns the (start, end) for music in the file"
        return self.startbyte, self.endbyte

    @property
    @memento
    def music_size(self):
        "Returns the total count of music data bytes in the file"
        return self.filesize - self.id3v1_totalsize - self.id3v2_totalsize

    def hash(self, hasher, maxbytes=None):
        """Returns the hash for a certain audio file ignoring tags """
        start, end = self.music_limits
        return hashfile(self.file, start, end, hasher, maxbytes)
